### ISSUE TITLE
Auto indent support

### ISSUE THREAD

**Issue by jonom** [adrhumphreys/vscode-silverstripe#13]
Thanks for this extension! It seems to work pretty well although I'm finding some of the things I can do with an html file don't work with a SS file. One example is auto indent... I like it when VSCode indents my lines automatically when I move them around. It seems a definition could be added to this extension to make that work, here is the info: https://code.visualstudio.com/updates/v1_14#_auto-indent-on-type-move-lines-and-paste

I'm not really following what is going on there but maybe it means something to you. I guess the indent rules could be copied from the html definition?

**Comment by jinjie** [comment]
Have you found a solution to this yet?

Installing this breaks many things that work with html syntax, including auto-indenting and emmet. I have to put these into my settings.json to have emmet working properly:

```json
{
    "editor.tabCompletion": "on",
    "editor.snippetSuggestions": "top"
}
```

**Comment by jonom** [comment]
No but I've been playing around with an idea this morning. I'm thinking that `.ss` files are really just html files with a little bit of extra markup. Any extensions and functions that are available in vsCode for HTML you would want to be available for `.ss` files as well, but by treating `.ss` files as a separate language that doesn't happen (as far as I can see). So I think that rather than define a whole language for `.ss` files, it makes more sense to just use the base HTML language, but extend it a little bit.

I followed [this guide on injecting grammars](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#injection-grammars) and was able to get some basic syntax highlighting happening with the following:

```json
{
	"scopeName": "ss-template.injection",
	"injectionSelector": "L:text.html",
	"patterns": [
		{
			"include": "#ss-template-comment"
		},
		{
			"include": "#ss-template-tag"
		}
	],
	"repository": {
		"ss-template-comment": {
			"match": "<%--.*?--%>",
			"name": "comment.sscomment"
		},
		"ss-template-tag": {
			"match": "<%.*?%>",
			"name": "keyword.sstag"
		}
	}
}
```

Maybe exploring that approach, and associating `.ss` files with HTML language would be a good solution?

**Comment by gorriecoe** [comment]
@jonom Let me know if the latest update helps.

**Comment by jonom** [comment]
Thanks for all your work @gorriecoe! It looks like auto indenting works but gets confused by SS tags. Might be partially a stylistic thing.

![indent](https://user-images.githubusercontent.com/1079425/87576266-f31dfc00-c685-11ea-9ca4-f82c1810508b.gif)

Also, not sure if it would be related but VSCode doesn't appear to know how to format .ss files

![image](https://user-images.githubusercontent.com/1079425/87576336-0f219d80-c686-11ea-81da-b047f3fc864b.png)

**Comment by gorriecoe** [comment]
Its probably my regex pattern.  I have updated it to this
```
<%\\s+(?:if|else|loop|with|control|cache|uncache)
```
I would like to simplify this to `.*` but found it would try and auto indent things like `<% include %>`.

As for the the formatter I'm not really sure.  Could be your settings as I changed the language name from uppercase to lowercase.  However I kept uppercase as an alias.  https://github.com/adrhumphreys/vscode-silverstripe/blob/master/package.json#L50-L54

**Comment by jonom** [comment]
So [this feature](https://code.visualstudio.com/docs/editor/codebasics#_formatting) works for you on `.ss` files?

I guess formatting might be too much to hope for since you can put SS tags *anywhere*... I can't imagine how you'd get perfect syntax highlighting or formatting when we can do things like this 😄

![image](https://user-images.githubusercontent.com/1079425/87597321-764f4a00-c6a6-11ea-9d02-3183fc7edee8.png)

**Comment by gorriecoe** [comment]
No that feature doesn't work for me.  =(  Was it previously?  I can investigate further.

For syntax highlighting its the same issue as https://github.com/adrhumphreys/vscode-silverstripe/issues/12 and is inherited from html.  I can't figure out how to resolve this both via or injection and include.  Otherwise we would have to write an entire .ss syntax

**Comment by jonom** [comment]
@gorriecoe no it wasn't working before either so no problem there.

**Comment by gorriecoe** [comment]
@jonom The latest update should have resolved the indentation of cache and uncached

**Comment by jonom** [comment]
@gorriecoe seems to be working a lot better - thanks! I noticed that `<h(x)>` tags seem to cause things below them to be indented... in case you're a sucker for pain and want to look at that too 😆  Certainly no expectation. Huge thank you for your improvements so far.

![indent2](https://user-images.githubusercontent.com/1079425/87971107-ac5f4600-ca79-11ea-9424-2bc91f46e17b.gif)

**Comment by gorriecoe** [comment]
@jonom I would say its an issue with the closing tag being on the same line.  The problem is a conflict of increase and decrease indention patterns.  The increase of indent rule kicks in after the regex matches but the decrease before the regex match.  If try that with any html element this will occur.  I have come across something similar and think I can fix this.  Just a matter of checking and excluding any html closing tag that is on the same line from increasing its indentation.  But yeah if you could confirm that is the issue that would be great.

**Comment by jonom** [comment]
@gorriecoe interesting - if I change the heading tag to a div (still all on one line) you're right, the same problem happens. Oddly though, if I put the heading opening and closing tags on their own lines, the issue still happens, even though it doesn't happen for the same structure with a div.

**Comment by gorriecoe** [comment]
@jonom https://github.com/adrhumphreys/vscode-silverstripe/commit/5a034ca3384762399e4ddb362e742e2283244a68.  The latest update should resolve both the h1 and sameline issues.  I'll get @adrhumphreys to publish it soon =)

**Comment by gorriecoe** [comment]
Published 1.0.4 should have this fixed :)

**Comment by jonom** [comment]
Legend! Thanks mate.
